exercises/Daniel_Murphy/ex01_basics/src/main.cpp

```diff
@@ -0,0 +1,151 @@
+#include <iostream>
+#include <fstream>
+#include <map>
+#include <string>
+#include <vector>
+#include <algorithm>
+#include <functional>
+using namespace std;
+
+class ReadFile
+{
+public:
+	std::vector<string> AnalyseLine(string line);
+	std::map<string, int> UpdateMap(std::vector<string> words);
+};
+
+
+std::vector<string> ReadFile::AnalyseLine(string line)
+{
+	std::vector<int> positions = {-1}; // holds positions of all spaces, include the start of the line
+
+	// Remove punctuation that accompanies a space
+	std::vector<char> punctuation = {'?','.',',', '\'', '"', '(', ')', '!', ':', ';'};
+	for (char mark : punctuation)
+	{
+		line.erase(std::remove(line.begin(), line.end(),mark), line.end());
+	}
+
+	// Hyphens don't accompany spaces so replace with a space and split words
+	std::replace( line.begin(), line.end(), '-', ' ');
+
+	// force all letters to case lower
+	std::transform(line.begin(), line.end(), line.begin(), [](unsigned char c){ return std::tolower(c);});
+
+	// Find position of space
+	int pos = static_cast<int>(line.find(" ",0));
+	while(pos != static_cast<int>(string::npos))
+	{
+		positions.push_back(pos);
+		pos = static_cast<int>(line.find(" ",pos+1));
+	}
+	positions.push_back(static_cast<int>(line.length())); // add the end of the line
+
+	// Cut the line by the space positions into words
+	std::vector<string> words;
+	for (auto iter = positions.begin(); iter != positions.end()-1; iter++)
+	{
+		int length = *(iter+1) - ((*iter)+1); // word length = end - start
+		// do not log spaces as words
+		if (length > 4)	{	words.push_back(line.substr(((*iter)+1), length));	}
+	}
```

Comment:
This chunk of code between L36 and L51 works fine but it can be done more concisely using the std::istringstream class (input stream). This mimics someone keying in data to a terminal via cin. Probably quickest to explain with an example. Something like this will do the equivalent work where the >> operator does the chopping up into words for (almost) free:
```
  std::istringstream lineOfText(line);
  std::string word;
  while (lineOfText >> word) {
    if (word.size() > 4)
      words.push_back(word);
  }
```
You'll need #include <sstream> also

Reply:
This is quite nice. I'd neglected sstream after happily using fstream already. Out of interest, on page 204, section 37.2 on string streams, do you think they meant option a. to be `std::istringstream` rather than a. and c. being the same??

Reply:
Yes you're right. a) should be std::istringstream